File: shared/app-state.ts
// Shared application state for the desktop app.
//
// The Deno/tray process owns this state and persists it to disk so polling can
// continue with the window closed. The webview is a view over the same state,
// reading and writing it through `/api/state` (see `server.ts`). The headless
// `watcher/` remains independent; this is the desktop app's own state.

import {
  type BrandDirectory,
  EMPTY_BRAND_DIRECTORY,
} from "../watcher/brands/directory.ts";
import {
  FUEL_CODES,
  type FuelCode,
  type FuelState,
  type FuelTrackerState,
  type StationSnapshot,
} from "../watcher/domain/model.ts";

export type SortMode = "distance" | "price";

export interface AppFilters {
  readonly longitude: number;
  readonly latitude: number;
  readonly radiusKm: number;
  readonly freeText: string;
  readonly selectedBrands: readonly string[];
  readonly selectedFuel: FuelCode | null;
  readonly onlyAvailable: boolean;
  readonly sortBy: SortMode;
  readonly notificationsEnabled: boolean;
}

export const DEFAULT_FILTERS: AppFilters = {
  longitude: 7.687329,
  latitude: 48.579682,
  radiusKm: 10,
  freeText: "",
  selectedBrands: [],
  selectedFuel: null,
  onlyAvailable: false,
  sortBy: "distance",
  notificationsEnabled: false,
};

export interface AppState {
  readonly version: 1;
  readonly filters: AppFilters;
  readonly brandDirectory: BrandDirectory;
  readonly brandArea: string | null;
  readonly trackers: Record<string, FuelTrackerState>;
  readonly updatedAt: string;
}

export function createInitialState(): AppState {
  return {
    version: 1,
    filters: DEFAULT_FILTERS,
    brandDirectory: EMPTY_BRAND_DIRECTORY,
    brandArea: null,
    trackers: {},
    updatedAt: new Date(0).toISOString(),
  };
}

function finiteOr(value: unknown, fallback: number): number {
  const parsed = Number(value);
  return Number.isFinite(parsed) ? parsed : fallback;
}

function positiveOr(value: unknown, fallback: number): number {
  const parsed = Number(value);
  return Number.isFinite(parsed) && parsed > 0 ? parsed : fallback;
}

function stringOr(value: unknown, fallback: string): string {
  return typeof value === "string" ? value : fallback;
}

function toFuelCode(value: unknown): FuelCode | null {
  if (typeof value !== "string") return null;
  return (FUEL_CODES as readonly string[]).includes(value)
    ? (value as FuelCode)
    : null;
}

export function normalizeFilters(input: unknown): AppFilters {
  const raw = (input ?? {}) as Partial<Record<keyof AppFilters, unknown>>;
  const brands = Array.isArray(raw.selectedBrands)
    ? raw.selectedBrands.filter((brand): brand is string =>
      typeof brand === "string"
    )
    : [];
  const sortBy = raw.sortBy === "price" || raw.sortBy === "distance"
    ? raw.sortBy
    : DEFAULT_FILTERS.sortBy;
  return {
    longitude: finiteOr(raw.longitude, DEFAULT_FILTERS.longitude),
    latitude: finiteOr(raw.latitude, DEFAULT_FILTERS.latitude),
    radiusKm: positiveOr(raw.radiusKm, DEFAULT_FILTERS.radiusKm),
    freeText: stringOr(raw.freeText, DEFAULT_FILTERS.freeText),
    selectedBrands: [...brands],
    selectedFuel: toFuelCode(raw.selectedFuel),
    onlyAvailable: raw.onlyAvailable === true,
    sortBy,
    notificationsEnabled: raw.notificationsEnabled === true,
  };
}

export function normalizeState(input: unknown): AppState {
  const raw = (input ?? {}) as Partial<Record<keyof AppState, unknown>>;
  const trackers = raw.trackers && typeof raw.trackers === "object"
    ? (raw.trackers as Record<string, FuelTrackerState>)
    : {};
  const directory = raw.brandDirectory &&
      typeof raw.brandDirectory === "object" &&
      typeof (raw.brandDirectory as BrandDirectory).updatedAt === "string" &&
      typeof (raw.brandDirectory as BrandDirectory).brands === "object"
    ? (raw.brandDirectory as BrandDirectory)
    : EMPTY_BRAND_DIRECTORY;
  return {
    version: 1,
    filters: normalizeFilters(raw.filters),
    brandDirectory: directory,
    brandArea: typeof raw.brandArea === "string" ? raw.brandArea : null,
    trackers,
    updatedAt: typeof raw.updatedAt === "string"
      ? raw.updatedAt
      : new Date(0).toISOString(),
  };
}

export interface SerializedStation extends Omit<StationSnapshot, "fuels"> {
  readonly fuels: Record<string, FuelState>;
}

export function serializeStation(station: StationSnapshot): SerializedStation {
  const fuels: Record<string, FuelState> = {};
  for (const [fuel, state] of station.fuels) fuels[fuel] = state;
  return { ...station, fuels };
}

export function deserializeStation(
  station: SerializedStation,
): StationSnapshot {
  const fuels = new Map<FuelCode, FuelState>();
  for (const [fuel, state] of Object.entries(station.fuels)) {
    fuels.set(fuel as FuelCode, state);
  }
  return { ...station, fuels };
}

export interface PollSnapshot {
  readonly stations: SerializedStation[];
  readonly fetchedAt: string;
  readonly dataUpdatedAt: string | null;
  readonly polling: boolean;
  readonly error: string | null;
  readonly pollIntervalMs: number;
}

export const EMPTY_POLL_SNAPSHOT: PollSnapshot = {
  stations: [],
  fetchedAt: new Date(0).toISOString(),
  dataUpdatedAt: null,
  polling: false,
  error: null,
  pollIntervalMs: 0,
};

File: src-tauri/src/app_state.rs
use serde::{Deserialize, Serialize};
use serde_json::Value;
use std::collections::BTreeMap;

use crate::brands::directory::{empty_brand_directory, BrandDirectory};
use crate::domain::model::{FuelCode, FuelTrackerState};

pub const EPOCH_ISO: &str = "1970-01-01T00:00:00.000Z";

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum SortMode {
    Distance,
    Price,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AppFilters {
    pub longitude: f64,
    pub latitude: f64,
    pub radius_km: f64,
    pub free_text: String,
    pub selected_brands: Vec<String>,
    pub selected_fuel: Option<FuelCode>,
    pub only_available: bool,
    pub sort_by: SortMode,
    pub notifications_enabled: bool,
}

impl Default for AppFilters {
    fn default() -> Self {
        AppFilters {
            longitude: 7.687329,
            latitude: 48.579682,
            radius_km: 10.0,
            free_text: String::new(),
            selected_brands: Vec::new(),
            selected_fuel: None,
            only_available: false,
            sort_by: SortMode::Distance,
            notifications_enabled: false,
        }
    }
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AppState {
    pub version: u32,
    pub filters: AppFilters,
    pub brand_directory: BrandDirectory,
    pub brand_area: Option<String>,
    pub trackers: BTreeMap<String, FuelTrackerState>,
    pub updated_at: String,
}

pub fn create_initial_state() -> AppState {
    AppState {
        version: 1,
        filters: AppFilters::default(),
        brand_directory: empty_brand_directory(),
        brand_area: None,
        trackers: BTreeMap::new(),
        updated_at: EPOCH_ISO.to_string(),
    }
}

fn to_number(value: &Value) -> Option<f64> {
    match value {
        Value::Number(number) => number.as_f64(),
        Value::String(text) => text.trim().parse::<f64>().ok(),
        Value::Bool(flag) => Some(if *flag { 1.0 } else { 0.0 }),
        Value::Null => Some(0.0),
        _ => None,
    }
}

fn finite_or(value: Option<&Value>, fallback: f64) -> f64 {
    value
        .and_then(to_number)
        .filter(|number| number.is_finite())
        .unwrap_or(fallback)
}

fn positive_or(value: Option<&Value>, fallback: f64) -> f64 {
    value
        .and_then(to_number)
        .filter(|number| number.is_finite() && *number > 0.0)
        .unwrap_or(fallback)
}

fn string_or(value: Option<&Value>, fallback: &str) -> String {
    match value {
        Some(Value::String(text)) => text.clone(),
        _ => fallback.to_string(),
    }
}

fn to_fuel_code(value: Option<&Value>) -> Option<FuelCode> {
    match value {
        Some(Value::String(text)) => FuelCode::parse(text),
        _ => None,
    }
}

pub fn normalize_filters(input: &Value) -> AppFilters {
    let defaults = AppFilters::default();
    let raw = input.as_object();
    let field = |key: &str| raw.and_then(|map| map.get(key));

    let selected_brands = match field("selectedBrands") {
        Some(Value::Array(items)) => items
            .iter()
            .filter_map(|item| item.as_str().map(|text| text.to_string()))
            .collect(),
        _ => Vec::new(),
    };

    let sort_by = match field("sortBy") {
        Some(Value::String(text)) if text == "price" => SortMode::Price,
        Some(Value::String(text)) if text == "distance" => SortMode::Distance,
        _ => defaults.sort_by,
    };

    AppFilters {
        longitude: finite_or(field("longitude"), defaults.longitude),
        latitude: finite_or(field("latitude"), defaults.latitude),
        radius_km: positive_or(field("radiusKm"), defaults.radius_km),
        free_text: string_or(field("freeText"), &defaults.free_text),
        selected_brands,
        selected_fuel: to_fuel_code(field("selectedFuel")),
        only_available: matches!(field("onlyAvailable"), Some(Value::Bool(true))),
        sort_by,
        notifications_enabled: matches!(
            field("notificationsEnabled"),
            Some(Value::Bool(true))
        ),
    }
}

pub fn normalize_state(input: &Value) -> AppState {
    let raw = input.as_object();

    let trackers = raw
        .and_then(|map| map.get("trackers"))
        .and_then(|value| serde_json::from_value::<BTreeMap<String, FuelTrackerState>>(value.clone()).ok())
        .unwrap_or_default();

    let brand_directory = raw
        .and_then(|map| map.get("brandDirectory"))
        .and_then(|value| serde_json::from_value::<BrandDirectory>(value.clone()).ok())
        .unwrap_or_else(empty_brand_directory);

    let brand_area = match raw.and_then(|map| map.get("brandArea")) {
        Some(Value::String(text)) => Some(text.clone()),
        _ => None,
    };

    let updated_at = match raw.and_then(|map| map.get("updatedAt")) {
        Some(Value::String(text)) => text.clone(),
        _ => EPOCH_ISO.to_string(),
    };

    AppState {
        version: 1,
        filters: normalize_filters(
            raw.and_then(|map| map.get("filters")).unwrap_or(&Value::Null),
        ),
        brand_directory,
        brand_area,
        trackers,
        updated_at,
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[test]
    fn normalize_filters_falls_back_to_defaults() {
        let filters = normalize_filters(&Value::Null);
        assert_eq!(filters.longitude, 7.687329);
        assert_eq!(filters.latitude, 48.579682);
        assert_eq!(filters.radius_km, 10.0);
        assert_eq!(filters.selected_fuel, None);
        assert!(!filters.notifications_enabled);
        assert!(filters.selected_brands.is_empty());
    }

    #[test]
    fn normalize_filters_coerces_numbers_and_drops_junk() {
        let filters = normalize_filters(&json!({
            "longitude": "3.5",
            "radiusKm": -1,
            "selectedFuel": "Gazole",
            "notificationsEnabled": true,
            "selectedBrands": ["A", 2, "B"]
        }));
        assert_eq!(filters.longitude, 3.5);
        assert_eq!(filters.radius_km, 10.0);
        assert_eq!(filters.selected_fuel, Some(FuelCode::Gazole));
        assert!(filters.notifications_enabled);
        assert_eq!(filters.selected_brands, vec!["A", "B"]);
    }

    #[test]
    fn normalize_state_rejects_unknown_fuels_and_bad_directories() {
        let state = normalize_state(&json!({
            "filters": { "selectedFuel": "Diesel" },
            "brandDirectory": { "brands": {} }
        }));
        assert_eq!(state.filters.selected_fuel, None);
        assert!(state.brand_directory.brands.is_empty());
        assert!(state.trackers.is_empty());
    }
}
